tls/extensions/s2n_server_certificate_status.c

```diff
@@ -23,64 +23,96 @@
 
 #define U24_SIZE 3
 
-static int s2n_server_certificate_status_send_size(struct s2n_connection *conn)
+/* In TLS 1.3, a response to a Status Request extension is sent as an extension with
+ * status request as well as the OCSP response. This contrasts to TLS 1.2 where
+ * the OCSP response is sent in the Certificate Status handshake message */
+
+static bool s2n_tls13_server_status_request_should_send(struct s2n_connection *conn);
+
+const s2n_extension_type s2n_tls13_server_status_request_extension = {
+    .iana_value = TLS_EXTENSION_STATUS_REQUEST,
+    .is_response = true,
+    .send = s2n_server_certificate_status_send,
+    .recv = s2n_server_certificate_status_parse,
+    .should_send = s2n_tls13_server_status_request_should_send,
+    .if_missing = s2n_extension_noop_if_missing,
+};
+
+static bool s2n_tls13_server_status_request_should_send(struct s2n_connection *conn)
+{
+    return s2n_server_can_send_ocsp(conn);
+}
+
+int s2n_server_certificate_status_send(struct s2n_connection *conn, struct s2n_stuffer *out)
 {
     notnull_check(conn);
-    if (s2n_server_can_send_ocsp(conn)) {
-        return sizeof(uint8_t) + U24_SIZE + conn->handshake_params.our_chain_and_key->ocsp_status.size;
-    }
+    struct s2n_blob *ocsp_status = &conn->handshake_params.our_chain_and_key->ocsp_status;
+    notnull_check(ocsp_status);
+
+    GUARD(s2n_stuffer_write_uint8(out, (uint8_t) S2N_STATUS_REQUEST_OCSP));
+    GUARD(s2n_stuffer_write_uint24(out, ocsp_status->size));
+    GUARD(s2n_stuffer_write(out, ocsp_status));
 
     return 0;
 }
 
-int s2n_tls13_ocsp_extension_send_size(struct s2n_connection *conn)
+int s2n_server_certificate_status_parse(struct s2n_connection *conn, struct s2n_stuffer *in)
 {
     notnull_check(conn);
-    if (s2n_server_can_send_ocsp(conn)) {
-        uint16_t size = 2 * sizeof(uint16_t);
-        GUARD_UINT16_AND_INCREMENT(s2n_server_certificate_status_send_size(conn), size);
-        return size;
+
+    uint8_t type;
+    GUARD(s2n_stuffer_read_uint8(in, &type));
+    if (type != S2N_STATUS_REQUEST_OCSP ) {
+        /* We only support OCSP */
+        return S2N_SUCCESS;
```

Comment:
should it error instead silently failing?

Reply:
Why would we want to error? If TLS introduced status request types other than OCSP, we wouldn't error on receiving them. We would just ignore them.